File: base/tests/test_OperationMultipleEval.py
# Copyright (C) 2008-today The SG++ project
# This file is part of the SG++ project. For conditions of distribution and
# use, please see the copyright notice provided with SG++ or at 
# sgpp.sparsegrids.org

import unittest

#correct the syspath, so python looks for packages in the root directory of SGpp
import sys, os
pathname = os.path.dirname(__file__)
pathlocal = os.path.abspath(pathname)
if pathlocal not in sys.path: sys.path.append(pathlocal)
pathsgpp = os.path.abspath(pathname) + '/tbin/tlearner'
if pathsgpp not in sys.path: sys.path.append(pathsgpp)


##
# Class with unittests for OperationMultipleEval
#
# @ingroup tests
#
class TestOperationMultipleEval(unittest.TestCase):

    ## 
    # Set up, create random DataVector and corresponding Python data structures.
    # @test DataVector::get(), DataVector::set()
    def setUp(self):
        pass

    ##
    # Constructors4.
    # @test DataVector::DataVector(size_t size), DataVector::DataVector(size_t size, size_t dim), DataVector::DataVector(DataVectorDefinition &DataVectorDef), DataVector::getSize(), DataVector::getDim(), DataVector::getSize()
    # @todo (pflueged) DataVector::DataVector(double *input, size_t size, size_t dim)
    def testConstructor(self):
        pass

    ##
    # Min, Max operations.
    # @test DataVector::min(int d), DataVector::max(int d), DataVector::minmax(int d, double *min, double *max), DataVector::min(), DataVector::max()
    def testOperationMultipleEval(self):
        from pysgpp import Grid
        from pysgpp import DataVector
        from pysgpp import DataMatrix
        from pysgpp import createOperationMultipleEval
        
        dim = 2
        level = 2
        grid = Grid.createLinearGrid(dim)
        grid.createGridGenerator().regular(level)
        gS = grid.getStorage()
        N = gS.size()
        alpha = DataVector([i + 1 for i in range(N)])
        
        points = [[0.5, 0.1], [0.3, 0.4], [0.9, 0.7]]
        
        numberDataPoints = 3
        
        result = DataVector([i for i in range(numberDataPoints)])
        
        dataset = DataMatrix(numberDataPoints, dim)
        for i in range(numberDataPoints):
            temp = DataVector(dim)
            for j in range(dim):
                print "p: ", points[i][j]
                temp[j] = points[i][j]
            dataset.setRow(i, temp)

        implementation = 7
        multiEvalOp = createOperationMultipleEval(grid, dataset, implementation)
        
        multiEvalOp.mult(alpha, result)        
        print 
        print "N: ", N
        print [alpha[i] for i in range(len(alpha))]
        print [result[i] for i in range(len(result))]
        #self.failUnlessEqual(quadOp.doQuadrature(alpha), qres)

# Run tests for this file if executed as application 
if __name__=='__main__':
    unittest.main()

File: base/tests/test_OperationMultipleEval.cpp
// Copyright (C) 2008-today The SG++ project
// This file is part of the SG++ project. For conditions of distribution and
// use, please see the copyright notice provided with SG++ or at
// sgpp.sparsegrids.org

#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>

#include <sgpp/base/grid/Grid.hpp>
#include <sgpp/base/datatypes/DataVector.hpp>
#include <sgpp/base/datatypes/DataMatrix.hpp>
// #include <sgpp/datadriven/DatadrivenOpFactory.hpp>
#include <sgpp/base/operation/BaseOpFactory.hpp>

using sgpp::base::BoundingBox1D;
using sgpp::base::DataMatrix;
using sgpp::base::DataVector;
using sgpp::base::Grid;
using sgpp::base::GridStorage;
using sgpp::base::OperationMultipleEval;

BOOST_AUTO_TEST_SUITE(TestOperationMultipleEval)

BOOST_AUTO_TEST_CASE(testOperationMultipleEval) {
  const size_t dim = 2;
  std::unique_ptr<Grid> grid(Grid::createLinearGrid(dim));
  grid->getGenerator().regular(2);

  grid->getBoundingBox().setBoundary(0, BoundingBox1D(3.0, 5.0));
  grid->getBoundingBox().setBoundary(1, BoundingBox1D(-2.0, 2.0));

  GridStorage& gS = grid->getStorage();

  size_t N = gS.getSize();

  DataVector alpha(N);

  for (int i = 0; i < static_cast<int>(N); ++i) {
    alpha[i] = static_cast<double>(i + 1);
  }

  // transformed to unit cube: {{0.5, 1.0}, {0.3, 0.4}, {0.9, 0.7}};
  const double points[3][2] = {{4.0, 2.0}, {3.6, -0.4}, {4.8, 0.8}};
  const size_t numberDataPoints = 3;

  DataVector result(numberDataPoints);

  for (unsigned int i = 0; i < (numberDataPoints); ++i) {
    result[i] = static_cast<double>(i);
  }

  DataMatrix dataset(numberDataPoints, dim);

  for (unsigned int i = 0; i < (numberDataPoints); ++i) {
    DataVector temp(dim);

    for (int j = 0; j < static_cast<int>(dim); ++j) {
      temp[j] = points[i][j];
    }

    dataset.setRow(i, temp);
  }

  sgpp::op_factory::createOperationMultipleEval(*grid, dataset)->mult(alpha, result);

  BOOST_TEST_MESSAGE(alpha.toString() + "\n");
  BOOST_TEST_MESSAGE(result.toString() + "\n");

  DataVector result_ref(numberDataPoints);
  result_ref[0] = 0.0;
  result_ref[1] = 2.72;
  result_ref[2] = 1.64;

  BOOST_CHECK_CLOSE(result[0], result_ref[0], 1e-7);
  BOOST_CHECK_CLOSE(result[1], result_ref[1], 1e-7);
  BOOST_CHECK_CLOSE(result[2], result_ref[2], 1e-7);
}

BOOST_AUTO_TEST_SUITE_END()
